process/throttle/block.go

```diff
@@ -0,0 +1,144 @@
+package throttle
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go/core"
+	"github.com/ElrondNetwork/elrond-go/core/logger"
+	"github.com/ElrondNetwork/elrond-go/process"
+)
+
+var log = logger.DefaultLogger()
+
+const jumpAbovePercent = 90
+const jumpBelowPercent = 90
+const jumpAboveFactor = 0.5
+const jumpBelowFactor = 0.5
+
+type blockInfo struct {
+	succeed  bool
+	round    uint64
+	items    uint32
+	maxItems uint32
+}
+
+// blockSizeThrottle implements BlockSizeThrottler interface which throttle block size
+type blockSizeThrottle struct {
+	statistics   []*blockInfo
+	maxItems     uint32
+	mutThrottler sync.RWMutex
+}
+
+// NewBlockSizeThrottle creates a new blockSizeThrottle object
+func NewBlockSizeThrottle() (*blockSizeThrottle, error) {
+	bst := blockSizeThrottle{}
+	bst.statistics = make([]*blockInfo, 0)
+	bst.maxItems = process.MaxItemsInBlock
+	return &bst, nil
+}
+
+// MaxItemsToAdd gets the maximum number of items which could be added in one block, taking into consideration the
+// previous results
+func (bst *blockSizeThrottle) MaxItemsToAdd() uint32 {
+	bst.mutThrottler.RLock()
+	maxItems := bst.maxItems
+	bst.mutThrottler.RUnlock()
+
+	return maxItems
+}
+
+// Add adds the new state for last block which has been sent and updates the max items which could be added
+// in one block
+func (bst *blockSizeThrottle) Add(round uint64, items uint32) {
+	bst.mutThrottler.Lock()
+	bst.statistics = append(
+		bst.statistics,
+		&blockInfo{round: round, items: items, maxItems: bst.maxItems},
+	)
+	bst.mutThrottler.Unlock()
+}
+
+// Succeed sets the state of the last block which has been sent at the given round
+func (bst *blockSizeThrottle) Succeed(round uint64) {
+	bst.mutThrottler.Lock()
+	for i := len(bst.statistics) - 1; i > 0; i-- {
```

Comment:
Shouldn't this be i>=0?

Reply:
Yep. I already fixed it when I have added units tests.